src/Strategy/Python/Poetry.hs

```diff
@@ -180,21 +187,68 @@ analyze PoetryProject{pyProjectToml, poetryLock} = do
 
 -- | Use a `pyproject.toml` to set the direct dependencies of a graph created from `poetry.lock`.
 setGraphDirectsFromPyproject :: Graphing Dependency -> PyProject -> Graphing Dependency
-setGraphDirectsFromPyproject graph pyproject = Graphing.promoteToDirect isDirect graph
+setGraphDirectsFromPyproject graph pyproject = graph'
   where
+    graph' :: Graphing Dependency
+    graph' = hydrateDepEnvs $ Graphing.promoteToDirect isDirect graph
+
+    directDeps :: [Dependency]
+    directDeps = pyProjectDeps pyproject
+
     -- Dependencies in `poetry.lock` are direct if they're specified in `pyproject.toml`.
     -- `pyproject.toml` may use non canonical naming, when naming dependencies.
     isDirect :: Dependency -> Bool
     isDirect dep = case pyprojectPoetry pyproject of
       Nothing -> False
-      Just _ -> any (\n -> toCanonicalName (dependencyName n) == toCanonicalName (dependencyName dep)) $ pyProjectDeps pyproject
+      Just _ -> any (\n -> toCanonicalName (dependencyName n) == toCanonicalName (dependencyName dep)) directDeps
 
 -- | Using a Poetry lockfile, build the graph of packages.
 -- The resulting graph contains edges, but does not distinguish between direct and deep dependencies,
 -- since `poetry.lock` does not indicate which dependencies are direct.
-graphFromLockFile :: PoetryLock -> Graphing Dependency
-graphFromLockFile poetryLock = Graphing.gmap pkgNameToDependency (edges <> Graphing.deeps pkgsNoDeps)
+graphFromPyProjectAndLockFile :: PyProject -> PoetryLock -> Graphing Dependency
+graphFromPyProjectAndLockFile pyProject poetryLock = graph
   where
+    -- Since Poetry lockfile v1.5 and gt, does not include category marker
+    -- it is not possible to know if the dependency is 'production' or 'development'.
+    -- strictly from lockfile itself. We hydrate "environment" from direct deps. If the
+    -- dependency has no environment, we mark it as dev environment. We know that all production
+    -- dependencies will be hydrated and will have some envionment.
+    graph :: Graphing Dependency
+    graph =
+      labelOptionalDepsIfPoetryGt1_5 $
+        hydrateDepEnvs $
+          Graphing.promoteToDirect isDirect $
+            Graphing.gmap pkgNameToDependency (edges <> Graphing.deeps pkgsNoDeps)
+
+    labelOptionalDepsIfPoetryGt1_5 :: Graphing Dependency -> Graphing Dependency
+    labelOptionalDepsIfPoetryGt1_5 g = if isLockLt1_5 then g else Graphing.gmap markEmptyEnvAsOptionalDep g
+
+    isLockLt1_5 :: Bool
+    isLockLt1_5 =
+      "0" `Text.isPrefixOf` lockVersion
+        || "1.0" `Text.isPrefixOf` lockVersion
+        || "1.1" `Text.isPrefixOf` lockVersion
+        || "1.2" `Text.isPrefixOf` lockVersion
+        || "1.3" `Text.isPrefixOf` lockVersion
+        || "1.4" `Text.isPrefixOf` lockVersion
```

Comment:
```suggestion
    isLockLt1_5 :: Bool
    isLockLt1_5 = any (`Text.isPrefixOf` lockVersion) ["0", "1.0", "1.1", "1.2", "1.3", "1.4"]
```

Reply:
I haven't tested that myself but you should be able to use a formulation like that ^^^